text/3476-expose-fn-type.md

```diff
@@ -0,0 +1,164 @@
+- Start Date: 2023-08-20
+- RFC PR: [rust-lang/rfcs#3476](https://github.com/rust-lang/rfcs/pull/3476)
+- Rust Issue: N/A
+
+# Summary
+[summary]: #summary
+
+This exposes the ghost-/inner-/localtype of a function to the user.
+
+# Motivation
+[motivation]: #motivation
+
+I was trying to make something similar to bevy's system functions. And for safety reasons, they check for conflicts between SystemParams, so that a function requiring `Res<A>` and `ResMut<A>` [panic](https://github.com/bevyengine/bevy/blob/main/crates/bevy_ecs/src/system/system_param.rs#L421).
+
+Then after I heard about axum's [`#[debug_handler]`](https://docs.rs/axum/latest/axum/attr.debug_handler.html) I wanted to do something similar to my copy of bevy systems, so that I get compile time errors when there is a conflict. I wanted even more, I wanted to force the user to mark the function with a specific proc attribute macro in order to make it possible to pass it into my code and call itself a system.
+
+For that, I would need to mark the type behind the function, for example, with a trait.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+As we all know, you can refer to a struct by its name and for example implement a trait
+```rust
+struct Timmy;
+impl Person for Timmy {
+    fn greet() {
+        println!("Hey it's me, Timmy!");
+    }
+}
+```
+When we want to target a specific function for a trait implementation, we somehow need to get to the type behind it. That is being done with the `fn` keyword as follows
+```rust
+fn my_function() {}
+impl MyTrait for fn my_function {
+    /* ... */
+}
+```
+---
+For a better understanding, imagine you have a struct like this:
+```rust
+struct FnContainer<F: Fn()> {
+    inner: F,
+}
+fn goods() { }
+
+let contained_goods = FnContainer {
+    inner: goods
+};
+```
+Here, we make a `FnContainer` which can hold every function with the signature `() -> ()` via generics.
+But what about explicitly designing the `FnContainer` for a specific function, just like the compiler does when resolving the generics. This will work the same as with the trait impl from above:
+```rust
+struct GoodsContainer {
+    inner: fn goods,
+}
+fn goods() {}
+
+let contained_goods = GoodsContainer {
+    inner: goods,
+}
+```
+---
+A function with a more complex signature, like with parameters, modifiers or a return type, is still just referenced by its name, because it's already unique
+```rust
+async fn request_name(id: PersonID) -> String { .. }
+
+impl Requestable for fn request_name {
+    /* ... */
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+As described in the **Guide-level explanation**, with the syntax `fn <fn_path>`, we can reference the type behind the named function.
+
+When the function is for example in a different mod, it should be referenced by its path
+```rust
+mod sub {
+    fn sub_mod_fn() { .. }
+}
+impl fn sub::sub_mod_fn {
+    /* ... */
+}
```

Comment:
Are specific function pointers currently treated as a local type (terminology from [E0210](https://github.com/rust-lang/rust/blob/master/compiler/rustc_error_codes/src/error_codes/E0210.md))? While this is definitely cool, the ability to add custom inherent impls on a specific function pointer brings a superset of possible issues that could be brought by normal trait impls. Should we analyze the two issues separately?

It is definitely [currently possible to implement a trait on function pointers in general](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=0dc8c5950decf2ea9b8f46ca672f440b), no matter specific or not, but I don't know if the inability to inherent-impl a specific fn pointer to a local function is merely due to its Voldemort properties or if it is intentionally so (or is it already possible through some trick I am not aware of?) or if it causes any incompatibility with the current dot-deref checks (since now `fn_name.foo()` could be `<fn fn_name>::foo(fn_name)`, `<&(fn fn_name>>::foo(fn_name)`, `<fn()>::foo(fn_name)`, `<&fn()>::foo(fn_name)`, etc).

Reply:
Yeah `<fn fn_name>::foo(fn_name)` should be possible... the other ones i kind of don't understand how they would be possible.

Reply:
Also ill add a note for similar behavior as E0210

Reply:
@DasLixou the `&` case is because of dot operator auto deref, the last two are because of type coercion

Reply:
But what does <fn()> here even mean? For me it looks like a Fn pointer